src/pie_utils/processor/document/regex_partitioner.py

```diff
@@ -0,0 +1,165 @@
+from __future__ import annotations
+
+import json
+import logging
+import re
+import statistics
+from typing import Any, Callable, Iterable, Iterator, Match
+
+from pytorch_ie.annotations import LabeledSpan
+
+from pie_utils.statistics import WithStatistics
+
+from ..document import DocumentWithPartition
+
+logger = logging.getLogger(__name__)
+
+
+def get_partitions_with_matcher(
+    document: DocumentWithPartition,
+    matcher: Callable[[str], Iterable[Match]],
+    label_group_id: int | None = None,  # = 1,
+    label_whitelist: list[str] | None = None,
+    skip_initial_partition: bool = False,  # = True
+    initial_partition_label: str = "initial part",
+) -> Iterator[LabeledSpan]:
+    """This method yields LabeledSpans as partitions of the given document. matcher is used to
+    search for a pattern in the document. If pattern is found then it returns Match objects that
+    contains matched groups. The required match group is selected using label_group_id to create a
+    partition label. The span of the partition ranges between two matched pattern. It should be
+    noted that none of the partitions overlap.
+
+    Besides the regular partitioning of the document, there are some other cases listed below:
+    1. Document can be converted into single partition with initial_partition_label if skip_initial_partition is False
+       and label_whitelist is empty list or label_group_id is None.
+    2. If only initial part of the document is required as a partition then set label_group_id and label_whitelist as
+       None and skip_initial_partition as False.
+    3. There will be no partitions if
+        a. matcher could not find pattern in document text
+        b. label_whitelist is empty list and skip_initial_partition is True
+        c. label_group_id and label_whitelist is None and skip_initial_partition is True
+        d. label_group_id is None and skip_initial_partition is False
+
+    :param document: A Document that is to be partitioned
+    :param matcher: A method that is used to match a pattern in the document text and return an iterator yielding the
+                    Match objects.
+    :param label_whitelist: A list of labels which are allowed to form a partition. (the default value is None)
+    :param label_group_id: An integer value to select suitable match group from Match object. (the default value is None)
+    :param skip_initial_partition: A boolean value that prevents initial partition to be saved in the document. (the
+                                default value is False)
+    :param initial_partition_label: A string value used as a partition label for initial partition. This is only used when
```

Comment:
missing default. general for defaults: you can just have "(default: value)" after it is first mentioned, e.g.: "A string value (default: "initial part") used as a  partition label ..."